src/components/TradeWidget/OrderValidity.tsx

```diff
@@ -58,137 +82,211 @@ const Wrapper = styled.div`
-  const { setValue, watch } = useFormContext<TradeFormData>()
-  // const error = errors[inputId]
-  const inputValue = watch(inputId)
-  // const overMax = ZERO
-  // const className = error ? 'error' : overMax.gt(ZERO) ? 'warning' : ''
-  // const errorOrWarning = error && <WalletDetail className="error">{error.message}</WalletDetail>
-
-  const handleChange = useCallback(() => {
-    const newValue = adjustPrecision(inputValue, 0)
-    if (inputValue !== newValue) {
-      setValue(inputId, newValue, true)
+const OrderValidity: React.FC<Props> = ({
+  isDisabled,
+  isUnlimited,
+  setAsap,
+  setUnlimited,
+  isAsap,
+  tabIndex,
+  validFromInputId,
+  validUntilInputId,
+}) => {
+  const [showOrderConfig, setShowOrderConfig] = useSafeState(false)
+  const { setValue, errors, register, getValues, watch } = useFormContext<TradeFormData>()
+  const { validFrom, validUntil } = getValues()
+
+  const validFromError = errors[validFromInputId]
+  const validUntilError = errors[validUntilInputId]
+  const validFromInputValue = watch(validFromInputId)
+  const validUntilInputValue = watch(validUntilInputId)
+  const overMax = ZERO
+  const validFromClassName = validFromError ? 'error' : overMax.gt(ZERO) ? 'warning' : ''
+  const validUntilClassName = validUntilError ? 'error' : overMax.gt(ZERO) ? 'warning' : ''
+
+  // const validFromErrorComp = validFromError && <WalletDetail className="error">{validFromError.message}</WalletDetail>
+  // const validUntilErrorComp = validUntilError && (
+  //   <WalletDetail className="error">{validUntilError.message}</WalletDetail>
+  // )
+
+  const handleShowConfig = useCallback((e): void => (e.preventDefault(), setShowOrderConfig(!showOrderConfig)), [
+    setShowOrderConfig,
+    showOrderConfig,
+  ])
+  const handleShowOrderConfig = useCallback((e): void => (e.preventDefault(), setShowOrderConfig(false)), [
+    setShowOrderConfig,
+  ])
+
+  const handleValidFromChange = useCallback(() => {
+    const newValue = adjustPrecision(validFromInputValue, 0)
+    if (validFromInputValue !== newValue) {
+      setValue(validFromInputId, newValue, true)
+    }
+  }, [validFromInputValue, setValue, validFromInputId])
+
+  const handleValidUntilChange = useCallback(() => {
+    const newValue = adjustPrecision(validUntilInputValue, 0)
+    if (validUntilInputValue !== newValue) {
+      setValue(validUntilInputId, newValue, true)
     }
-  }, [inputValue, setValue, inputId])
+  }, [validUntilInputValue, setValue, validUntilInputId])
 
   useEffect(() => {
-    handleChange()
-  }, [handleChange])
+    handleValidFromChange()
+  }, [handleValidFromChange])
 
-  // function handleUnlimitedClick(): void {
-  //   setUnlimited(!isUnlimited)
-  //   !isUnlimited ? setValue(inputId, '', true) : setValue(inputId, '30', true)
-  // }
+  useEffect(() => {
+    handleValidUntilChange()
+  }, [handleValidUntilChange])
+
+  function handleUnlimitedClick(): void {
+    setUnlimited(!isUnlimited)
```

Comment:
Shouldn't it be `setUnlimited(unlimited => !unlimited)` ? same for `setAsap`

Reply:
i think both are fine but ill go with this for safety @Velenir

Reply:
Both are fine here
function setState is used when we either don't have access to current state or to account for multiclicks or similar. As in
`setState(false)  x 10 times` will end up false, `setState(flip => !flip)  x 10 times` will end up the opposite